### PR TITLE
Fix A.fromList crashes on empty list of type [()]

### PR DESCRIPTION
**Pull request by rnjtranjan**

### CODE REVIEW

**Comment by rnjtranjan** [comment]
fixes #755

File: src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 1]
```diff
@@ -291,7 +291,7 @@ foreign import ccall unsafe "string.h memcmp" c_memcmp
 {-# INLINE bytesToElemCount #-}
 bytesToElemCount :: Storable a => a -> Int -> Int
 bytesToElemCount x n =
-    let elemSize = sizeOf x
+    let elemSize = max 1 $ sizeOf x
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:294]: When can `sizeOf` something be `0`?

Comment (rnjtranjan) [reply src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:294]: sizeOf  "()" is zero.

**Comment by harendra-kumar** [comment]
This is not completely fixed. We need to replace all places where `sizeOf` is being used. We can create an internal `sizeOfElem` wrapper over `sizeOf` and use that at all places wherever we are using `sizeOf`.

File: src/Streamly/Internal/System/IO.hs [hunk 2]
```diff
@@ -40,7 +40,7 @@ unsafeInlineIO (IO m) = case m realWorld# of (# _, r #) -> r
 -- See https://gitlab.haskell.org/ghc/ghc/-/wikis/commentary/rts/storage/heap-objects#arrays
 --
 byteArrayOverhead :: Int
-byteArrayOverhead = 2 * sizeOf (undefined :: Word)
+byteArrayOverhead = 2 * sizeOfElem (undefined :: Word)
```
Comment (harendra-kumar) [inline src/Streamly/Internal/System/IO.hs:43]: revert this.

File: test/Streamly/Test/Common/Array.hs [hunk 3]
```diff
@@ -83,7 +84,7 @@ maxTestCount = 10
 #endif
 
 allocOverhead :: Int
-allocOverhead = 2 * sizeOf (undefined :: Int)
+allocOverhead = 2 * sizeOfElem (undefined :: Int)
```
Comment (harendra-kumar) [inline test/Streamly/Test/Common/Array.hs:87]: revert this.

File: test/Streamly/Test/FileSystem/Handle.hs [hunk 4]
```diff
@@ -36,7 +37,7 @@ import Test.Hspec as H
 import Test.Hspec.QuickCheck
 
 allocOverhead :: Int
-allocOverhead = 2 * sizeOf (undefined :: Int)
+allocOverhead = 2 * sizeOfElem (undefined :: Int)
```
Comment (harendra-kumar) [inline test/Streamly/Test/FileSystem/Handle.hs:40]: revert this.

File: src/Streamly/Internal/Wrapper.hs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    ( sizeOfElem
+    , sizeOfPrimElem
+    )
+where
+
+import qualified Foreign.Storable  as ST
+import Data.Primitive.Types as Prim (Prim, sizeOf )
+
+sizeOfElem :: forall a . (ST.Storable a) => a -> Int
+sizeOfElem _ = max 1 $ ST.sizeOf (undefined :: a)
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Wrapper.hs:20]: We can define this in Data.Array.Foreign.Mut.Type. This is array specific and not a general API.

File: src/Streamly/Internal/Wrapper.hs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+sizeOfPrimElem :: forall a . (Prim a) => a -> Int
+sizeOfPrimElem _ = max 1 $ sizeOf (undefined :: a)
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Wrapper.hs:23]: We can define this in the lowest level prim array module.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

File: src/Streamly/Internal/Data/Array/Stream/Foreign.hs [hunk 7]
```diff
@@ -413,8 +411,7 @@ splitAtArrayListRev n ls
                 else if m == len
                 then ([x],xs)
                 else let !(Array contents start end) = x
-                         sz = sizeOfElem (undefined :: a)
-                         end1 = end `plusPtr` negate (m * sz)
+                         end1 = end `plusPtr` negate (m * SIZE_OF(a))
```
Comment (rnjtranjan) [inline src/Streamly/Internal/Data/Array/Stream/Foreign.hs:414]: May be we can create a macro for this

**Review by rnjtranjan — COMMENTED** [review]
looks good

**Pull request merged**